File: rust/examples/substrate_staking.rs
// Substrate Staking Example
// Demonstrates staking operations using the Selendra SDK

use selendra_sdk::substrate::{Connection, keypair_from_string};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Connect to Selendra testnet
    let conn = Connection::new("wss://rpc-testnet.selendra.org").await?;
    println!("✅ Connected to Selendra");

    // Create keypair from seed phrase
    let keypair = keypair_from_string("//Alice");
    println!("📝 Account: {}", keypair.to_ss58check());

    // Sign connection
    let signed = conn.sign(&keypair)?;
    println!("✅ Signed connection ready");

    // Query current balance
    let balance = signed.get_balance().await?;
    println!("💰 Balance: {} units", balance);

    // Query staking info
    let staking_info = conn.get_staking_info(keypair.account_id()).await?;
    println!("🎯 Staking Info:");
    println!("  - Bonded: {:?}", staking_info.bonded);
    println!("  - Min validators: {}", staking_info.minimum_validator_count);
    println!("  - Sessions per era: {}", staking_info.sessions_per_era);

    // Example: Bond tokens for staking
    // UNCOMMENT to execute (requires sufficient balance)
    // let amount = 1_000_000_000_000_000_000u128; // 1 SEL
    // let tx_hash = signed.stake_bond(amount).await?;
    // println!("✅ Bonded tokens: {}", tx_hash);

    // Example: Nominate validators
    // let nominees = vec![/* validator addresses */];
    // let tx_hash = signed.stake_nominate(nominees).await?;
    // println!("✅ Nominated validators: {}", tx_hash);

    // Example: Start validating (requires bonded stake)
    // let commission = 10; // 10% commission
    // let tx_hash = signed.stake_validate(commission).await?;
    // println!("✅ Started validating: {}", tx_hash);

    // Example: Bond extra tokens
    // let extra = 500_000_000_000_000_000u128; // 0.5 SEL
    // let tx_hash = signed.stake_bond_extra(extra).await?;
    // println!("✅ Bonded extra: {}", tx_hash);

    // Example: Stop nominating/validating
    // let tx_hash = signed.stake_chill().await?;
    // println!("✅ Chilled: {}", tx_hash);

    Ok(())
}

File: old/typescript/examples/substrate-staking.ts
/**
 * Substrate Staking Example
 * Demonstrates staking operations using StakingClient
 */

import { ApiPromise, WsProvider } from '@polkadot/api';
import { Keyring } from '@polkadot/keyring';
import { StakingClient } from '../src/substrate/staking';

async function main() {
  console.log('🔗 Connecting to Selendra...');

  // Connect to Selendra
  const provider = new WsProvider('wss://rpc-testnet.selendra.org');
  const api = await ApiPromise.create({ provider });
  await api.isReady;
  console.log('✅ Connected');

  // Initialize staking client
  const stakingClient = new StakingClient(api);

  // Create keyring
  const keyring = new Keyring({ type: 'sr25519' });
  const alice = keyring.addFromUri('//Alice');

  console.log('\n📝 Staking Account:', alice.address);

  // Query staking information
  console.log('\n🎯 Querying Staking Information...');
  try {
    const stakingInfo = await stakingClient.getStakingInfo(alice.address);
    if (stakingInfo) {
      console.log('  Total Stake:', stakingInfo.totalStake);
      console.log('  Active Stake:', stakingInfo.activeStake);
      console.log('  Own Stake:', stakingInfo.ownStake);
      console.log('  Era:', stakingInfo.era);
      console.log('  Session Index:', stakingInfo.sessionIndex);
    } else {
      console.log('  ⚠️  No staking info (account not staking yet)');
    }
  } catch (error) {
    console.log('  ⚠️  Error querying staking info');
  }

  // Example: Bond tokens (COMMENTED OUT)
  /*
  console.log('\n💎 Bonding tokens...');
  const bondAmount = '1000000000000000000'; // 1 SEL
  const bondResult = await stakingClient.bond(
    alice.address, // controller
    bondAmount,
    { signer: alice, waitForFinalization: true }
  );
  console.log('✅ Bonded:', bondResult.hash);
  */

  // Example: Nominate validators (COMMENTED OUT)
  /*
  console.log('\n🗳️  Nominating validators...');
  const validators = [
    '5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY',
    '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty',
  ];
  const nominateResult = await stakingClient.nominate(
    validators,
    { signer: alice, waitForFinalization: true }
  );
  console.log('✅ Nominated:', nominateResult.hash);
  */

  // Example: Become validator (COMMENTED OUT)
  /*
  console.log('\n⚡ Registering as validator...');
  const commission = 10; // 10%
  const validateResult = await stakingClient.validate(
    commission,
    { signer: alice, waitForFinalization: true }
  );
  console.log('✅ Validating:', validateResult.hash);
  */

  // Query validator information
  console.log('\n👥 Querying Validators...');
  const validators = await stakingClient.getValidators();
  console.log(`  Active validators: ${validators.length}`);
  if (validators.length > 0) {
    console.log(`  First validator: ${validators[0]}`);
  }

  // Disconnect
  await api.disconnect();
  console.log('\n👋 Disconnected');
}

main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error('❌ Error:', error);
    process.exit(1);
  });
